### PR TITLE
Release 2.0.0

### LINKED ISSUE #207: Address issues required to deploy gamification
- [x] Update Google Cloud SQL Database job times out after 50mins (sampledata)
- [x] Update version number
- [x] Create missed new migration
- [x] Up Django to 2.2

### PR DESCRIPTION
**Pull request by courtneycb**
**Changes**

- Add gamification system that allows users to earn points and badges.
- Add more filters for question attempts on the admin site.
- Add cron job that every minute will backdate any 15 user profiles that haven't already been backdated.
- Update Django from 2.1.5 to 2.2.3.
- Dependency updates:
    - Removed django-coverage-plugin.
    - Update coverage from 4.5.4 to 5.0.4.
    - Update coverage from 5.0.4 to 5.1.
    - Update django-activeurl from 0.1.12 to 0.2.0.
    - Update django-allauth from 0.39.1 to 0.41.0.
    - Update django-ckeditor from 5.7.1 to 5.9.0.
    - Update django-crispy-forms from 1.7.2 to 1.9.0.
    - Update django-debug-toolbar from 2.0 to 2.2.
    - Update django-extensions from 2.2.1 to 2.2.9.
    - Update django-modeltranslation from 0.13.3 to 0.14.4.
    - Update django-model-utils from 3.2.0 to 4.0.0.
    - Update django-recaptcha from 2.0.5 to 2.0.6.
    - Update django-redis from 4.10.0 to 4.11.0.
    - Update django-storages from 1.7.1 to 1.9.1.
    - Update google-api-python-client from 1.7.11 to 1.7.12.
    - Update google-auth from 1.6.3 to 1.12.0.
    - Update google-cloud-logging from 1.12.1 to 1.15.0.
    - Update mypy from 0.720 to 0.770.
    - Update Pillow from 6.1.0 to 7.0.0.
    - Update pydocstyle from 4.0.1 to 5.0.2.
    - Update pytest from 5.1.1 to 5.4.1.
    - Update pytest-django from 3.5.1 to 3.8.0.
    - Update PyYAML from 5.1.2 to 5.3.
    - Update Sphinx from 2.2.0 to 2.4.4.

Resolves #207

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,39 @@
 # Changelog
 
+## 1.6.0
+- Add gamification system that allows users to earn points and badges.
+- Add more filters for question attempts on the admin site.
+- Add cron job that every minute will backdate any 15 user profiles that haven't already been backdated.
```
Comment (eAlasdair) [inline CHANGELOG.md:6]: From top to here:
- This is the big one, so maybe emphasise it a little more
- This is not wrong, but it is inconsequentially small. Many more, more important things are added to the admin site so reword it as a more general `admin is improved` type report
- This doesn't mean anything to a user - cron? backdate? user profile? ¯\_(ツ)_/¯ Maybe something like "Add ability to recalculate a user's points and badges from their past attempts." and group it with the gamification report

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,36 @@
...
+- Update Django from 2.1.5 to 2.2.3.
+- Dependency updates:
+    - Removed django-coverage-plugin.
```
Comment (eAlasdair) [inline CHANGELOG.md:9]: Make present tense for consistency?

File: CHANGELOG.md [hunk 3]
```diff
@@ -37,7 +68,7 @@
 - Fix bug where whitespace of user code in attempt wasn't shown in admin interface.
 - Fix bug where sender's email address is not listed on contact us forms sent to admin.
```
Comment (eAlasdair) [inline CHANGELOG.md:69]: Unrelated but I noticed this is #53 (open issue), so should the issue be resolved?

File: codewof/templates/base.html [hunk 4]
```diff
@@ -52,11 +52,14 @@
...
           {% if request %}
             {% activeurl %}
-            <ul class="navbar-nav w-100 d-lg-flex justify-content-end align-items-stretch">
+            <ul id="codewof-navbar-dashboard" class="navbar-nav w-100 d-lg-flex justify-content-end align-items-stretch">
               {% if request and request.user.is_authenticated %}
                 <a class="nav-item nav-link mr-auto" href="{% url 'programming:question_list' %}">Questions</a>
                 {# URL provided by django-allauth/account/urls.py #}
                 <a class="nav-item nav-link" href="{% url 'users:dashboard'  %}">
+                  {% svg 'icons8-points' %}
+                  <span id="user_points_navbar">{{request.user.profile.points}}</span>
```
Comment (eAlasdair) [inline codewof/templates/base.html:61]: for consistency add a space before/after inside the curly braces

File: codewof/templates/users/achievements.html [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  <div class="container">
+    <h1 class="display-3 orange-underline text-center">Achievements</h1>
+    <h3>Achievements earned: {{num_badges_earned}}/{{num_badges}}</h3>
+    <div class="col-12 col-md-8 offset-md-2 mb-3">
+        <br>
+        <div class="row">
+            {% for badge in user.profile.earned_badges.all %}
+                <div class="col-4 badge-container">
+                    <img class="badge-icon" src="{% static badge.icon_name %}">
+                    <b>{{badge.display_name}}</b>
```
Comment (eAlasdair) [inline codewof/templates/users/achievements.html:16]: for consistency add a space before/after inside the curly braces (also on line 9)

File: codewof/templates/users/dashboard.html [hunk 6]
```diff
@@ -30,11 +30,23 @@ <h2>Your questions for today</h2>
             <strong>Sorry!</strong> We currently don't have questions for you to do today.
           </div>
         {% endif %}
+        <p>You've solved {{num_questions_answered}} question{{ num_questions_answered|pluralize }} in the last month!</p>
```
Comment (eAlasdair) [inline codewof/templates/users/dashboard.html:33]: for consistency add a space before/after inside the curly braces (Also line 37, 45)

File: codewof/templates/users/user_detail.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+{% block content %}
+  <div class="container">
+
+    <div class="row">
+      <div class="col-sm-12">
+        <h1>{{ user.get_full_name }}</h1>
+
+        <p><strong>Date joined</strong>: {{ user.date_joined }}</p>
+        <p><strong>Points earned</strong>: {{user.profile.points}}</p>
```
Comment (eAlasdair) [inline codewof/templates/users/user_detail.html:14]: for consistency add a space before/after inside the curly braces (also line 20, 22)

File: codewof/tests/programming/test_models.py [hunk 8]
```diff
@@ -1,153 +1,390 @@
...
+    #     Earned.objects.create(profile=user.profile, badge=badge)
+    #     check_badge_conditions(user.profile)
+
+    #     earned = Earned.objects.filter(profile=user.profile, badge=badge)
+    #     self.assertEqual(len(earned), 1)
+
+    # def test_adding_unknown_badge_doesnt_break(self):
+    #     Badge.objects.create(id_name="notrealbadge", display_name="test", description="test")
+    #     user = User.objects.get(pk=1)
+    #     check_badge_conditions(user.profile)
```
Comment (eAlasdair) [inline codewof/tests/programming/test_models.py:143]: Commented out tests? (also at end of file)

**Comment by eAlasdair** [comment]
@courtneycb if you're happy with these changes I'll approve the PR

File: codewof/templates/users/activity_graph.html [hunk 9]
```diff
@@ -0,0 +1,65 @@
+<!DOCTYPE html>
```
Comment (JackMorganNZ) [inline codewof/templates/users/activity_graph.html:1]: What is this file used for?

Comment (courtneycb) [reply codewof/templates/users/activity_graph.html:1]: This is related to #203 and should be removed from this PR.

File: codewof/templates/programming/badge_card.html [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{% load static %}
+
+<div class="qc-container">
+    <a>
+        <div class="qc-checkbox">
+            <img src="{% static badge.icon_name %}">
+        </div>
+        <div class="qc-icon">
+
+        </div>
```
[the remark this answers is not in the document]
Comment (courtneycb) [reply codewof/templates/programming/badge_card.html:10]: Unsure, removing.

**Comment by eAlasdair** [comment]
Some instances of 'badge' shouldn't have been renamed to 'achievement'

**Pull request merged**